src/peergos/server/storage/GarbageCollector.java

```diff
@@ -0,0 +1,148 @@
+package peergos.server.storage;
+
+import peergos.server.util.*;
+import peergos.shared.cbor.*;
+import peergos.shared.crypto.hash.*;
+import peergos.shared.io.ipfs.multihash.*;
+import peergos.shared.storage.*;
+
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.*;
+import java.util.logging.*;
+
+public class GarbageCollector implements ContentAddressedStorage {
+
+    private static final long MAX_WAIT_FOR_TRANSACTION_MILLIS = 10_000;
+
+    private final ContentAddressedStorage target;
+    private final long gcPeriodMillis;
+    // This lock is used to make new transactions block until a pending GC completes
+    private final Object gcLock = new Object();
+    private final ConcurrentHashMap<PublicKeyHash, AtomicInteger> openTransactions = new ConcurrentHashMap<>();
+
+    public GarbageCollector(ContentAddressedStorage target, long gcPeriodMillis) {
+        this.target = target;
+        this.gcPeriodMillis = gcPeriodMillis;
+    }
+
+    public void start() {
+        new Thread(this::run).start();
+    }
+
+    private int openTransactions() {
+        int res = 0;
+        for (AtomicInteger open : openTransactions.values()) {
+            res += Math.max(0, open.get());
+        }
+        return res;
+    }
+
+    public void run() {
+        while (true) {
+            try {
+                synchronized (gcLock) {
+                    long start = System.nanoTime();
+                    while (openTransactions() > 0) {
+                        if ((System.nanoTime() - start) / 1_000_000 > MAX_WAIT_FOR_TRANSACTION_MILLIS)
+                            openTransactions.clear();
+                        System.out.println("GC sleeping, " + openTransactions() + " open transactions");
+                        Thread.sleep(100);
+                    }
+                    long ready = System.nanoTime();
+                    target.gc().join();
+                    long done = System.nanoTime();
+                    Logging.LOG().info(String.format("GC took: %d ms waiting to start, %d ms in actual GC",
+                            (ready - start)/1000000, (done - ready)/1000000));
```

Comment:
add underscores to 1000000 for readability

Reply:
done